### DISCUSSION TITLE
Support for foreign joins with GetForeignJoinPaths

### DISCUSSION THREAD

**Discussion by fungiboletus** [supabase/supabase#26630] [Contribute to Supabase]
Hi,

I'm playing with pgrx and wrappers and it's great!

I understood that in the current implementation, PostgreSQL will fetch all the foreign data and compute a join locally. I see that PostgreSQL supports [GetForeignJoinPaths](https://www.postgresql.org/docs/current/fdw-callbacks.html#FDW-CALLBACKS-JOIN-SCAN) to have foreign joins, that should greatly improve performances for my uses cases.

Is there any plan to support foreign joins in supabase/wrappers? If not, how hard do you think it is to implement that in a pull request? Is there major roadblocks or it's not there because it hasn't been needed so far?

**Comment by burmecia** [comment]
[GetForeignJoinPaths](https://www.postgresql.org/docs/current/fdw-callbacks.html#FDW-CALLBACKS-JOIN-SCAN) isn't supported yet because of its complexity and not all sources support joins. Can you describe more details about your use cases? If you need to join tables remotely, maybe can consider this [remote subquery feature](https://supabase.github.io/wrappers/clickhouse/#foreign-table-options)? Note it's available to RDBMS sources only. By using that feature, you can create a foreign table just reading the result of remote table joins. For example,

```sql
create foreign table my_foreign_table (
  id bigint,
  name text
)
  server clickhouse_server
  options (
    table '(select * from my_table a join another_table b on a.id = b.id)'
  );
```

**Reply by fungiboletus** [reply]
I realise that `GetForeignJoinPaths` is perhaps not what I need.

Here is one (simple) example use case. I havn't ran it so it may not be perfect, but it should illustrate what I want to achieve.

In PostgreSQL:

```sql
CREATE TABLE airplane (
    id UUID PRIMARY KEY,
    brand TEXT,
    sign TEXT
);
```

In ClickHouse:
```sql
CREATE TABLE airplane_position (
    airplane_uuid UUID,
    lat Float64,
    long Float64,
    altitude Float64,
    datetime DateTime64(3)
) ENGINE = MergeTree
ORDER BY (datetime, airplane_uuid);
```

And the corresponding FDW in PostGresql:
```sql
CREATE FOREIGN TABLE clickhouse_airplane_position (
    airplane_uuid uuid,
    lat double precision,
    long double precision,
    altitude double precision,
    datetime timestamp
)
SERVER clickhouse_server
OPTIONS (
    table 'airplane_position'
);
```

If I understood correctly, the following kind of queries works well and will fetch the data in ClickHouse with the right *quals*:
```sql
SELECT *
FROM clickhouse_airplane_position
WHERE airplane_uuid IN (
    '123e4567-e89b-12d3-a456-426614174000',
    '123e4567-e89b-12d3-a456-426614174001', 
    '123e4567-e89b-12d3-a456-426614174002' 
);
```

But a query with a join will fetch all the remote data from clickhouse and execute the join locally. I don't want that, as I may have many airplanes.

```sql
SELECT airplane.id, airplane.brand, airplane.sign, position.lat, position.long, position.altitude, position.datetime
FROM airplane
JOIN clickhouse_airplane_position AS position ON airplane.id = position.airplane_uuid
WHERE airplane.id IN (
    '123e4567-e89b-12d3-a456-426614174000', 
    '123e4567-e89b-12d3-a456-426614174001',
    '123e4567-e89b-12d3-a456-426614174002' 
);
```

Moreover, the PostgreSQL's query optimiser is good at detecting joins and writing a query like the first one will still be executed as a join.

This query will be a join, and fetch all the foreign data from clickhouse:

```sql
SELECT *
FROM clickhouse_airplane_position
WHERE airplane_uuid IN (
  SELECT airplane.id FROM airplane WHERE airplane.brand = 'Sud Aviation'
);
```

I found a hacky solution to not trigger a join by using some dynamic SQL:

```sql
CREATE OR REPLACE FUNCTION plz_no_join(_ids uuid[])
   RETURNS TABLE(id uuid, lat double precision, long double precision, altitude double precision, datetime timestamp) AS
$func$
BEGIN
   RETURN QUERY EXECUTE
     'SELECT *
      FROM clickhouse_airplane_position
      WHERE  airplane_uuid = ANY ($1)'
   USING _ids);
END
$func$  LANGUAGE plpgsql;

SELECT *
FROM plz_no_join(array(SELECT airplane.id FROM airplane WHERE airplane.brand = 'Sud Aviation'));
```

Ideally, I would like to be able to join my two foreign table with the local table without having to write exotic SQL. But it may not be possible.

**Reply by burmecia** [reply]
Thanks for the detailed example. Joining on foreign table and doing correct pushdown is challenging, it might work well in some simple SQLs but covering all cases is a hard task. I think ELT (Extraction, Load and Transform) is better approach to use FDW and foreign tables.

For you example, IMO a better pattern is firstly loading the `clickhouse_airplane_position` table data to local Postgres db, and then doing whatever joins or processing for free because the data is already in local db. If you need keep local and remote in sync, you can set up a job, for example using `pg_cron`, to read the data from foreign table to a local table regularly.

So I think FDW and foreign table is more suitable to be used as data extraction tool rather than a normal PG table, as the joins and pushdown may not work efficiently in some SQLs.

**Reply by fungiboletus** [reply]
I understand. I liked the idea of having a quick foreign table with always the latest data, but this is perhaps a wild dream with PostgreSQL's foreign data wrappers currently. Having copies of the data in postgresql is a good suggestion for cases where it's not too much data.

I will also play a bit with [Presto](https://github.com/prestodb/presto) to see whether it handles this kind of joins. It has a huge overhead and it involves the JVM and JDBC drivers, but it may do what I want.
